File: Tests/Review3/prim/prim.py
from typing import List


def prim(edges: List[List[int]], points: int) -> tuple[int, List[List[int]]]:
    path: List[List[int]] = []
    cost = 0

    count = points - 1
    visited: List[bool] = [False] * points
    queue: List[List[int]] = []

    # init
    visited[0] = True
    for item in edges:
        if item[0] == 0 or item[1] == 0:
            queue.append(item)

    while count > 0:
        queue.sort(key=lambda x: x[2])
        v, w, weight = queue.pop(0)

        if not visited[v] or not visited[w]:
            cost += weight
            path.append([v, w])
            count -= 1

            next_point = v if not visited[v] else w
            visited[next_point] = True

            for item in edges:
                if (item[0] == next_point and not visited[item[1]]) or (
                    item[1] == next_point and not visited[item[0]]
                ):
                    queue.append(item)

    return cost, path


# test
res = prim(
    [
        [0, 1, 4],
        [0, 2, 13],
        [0, 3, 7],
        [0, 4, 7],
        [1, 2, 9],
        [1, 3, 3],
        [1, 4, 7],
        [2, 3, 10],
        [2, 4, 14],
        [3, 4, 4],
    ],
    5,
)
print(res)

File: Data Structure & Algorithm/Review/Stage 1/Review 3/prim/prim.ts
export function prim(graph: number[][], vertices: number): [number, number[][]] {
  let cost = 0;
  const path: number[][] = [];

  const visited: boolean[] = new Array(vertices).fill(false);
  let count = vertices - 1;
  const queue: number[][] = [];

  // init
  visited[0] = true;
  for (const item of graph) {
    if (item[0] === 0 || item[1] === 0) {
      queue.push(item);
    }
  }

  while (count > 0 && queue.length) {
    queue.sort((a, b) => a[2] - b[2]);
    const [v, w, weight] = queue.shift()!;
    if (!visited[v] || !visited[w]) {
      const new_node = !visited[v] ? v : w;
      cost += weight;
      path.push([v, w]);
      count--;
      visited[new_node] = true;

      for (const item of graph) {
        if ((item[0] === new_node && !visited[item[1]]) || (item[1] === new_node && !visited[item[0]])) {
          queue.push(item);
        }
      }
    }
  }

  return [cost, path];
}

// test
const res = prim(
  [
    [0, 1, 4],
    [0, 2, 13],
    [0, 3, 7],
    [0, 4, 7],
    [1, 2, 9],
    [1, 3, 3],
    [1, 4, 7],
    [2, 3, 10],
    [2, 4, 14],
    [3, 4, 4],
  ],
  5
);
console.log(res);
